File: solutions/typescript/problems/[1101_Medium] The Earliest Moment When Everyone Become Friends.ts
class DisjointSetWithCount {
    private root: number[] = [];
    private rank: number[] = [];
    private count: number;

    constructor(size: number) {
        this.count = size;
        for (let i = 0; i < size; i++) {
            this.root[i] = i;
            this.rank[i] = 1;
        }
    }

    find(x: number): number {
        if (x === this.root[x]) {
            return x;
        }
        return (this.root[x] = this.find(this.root[x]));
    }

    union(x: number, y: number): void {
        const rootX = this.find(x);
        const rootY = this.find(y);
        if (rootX !== rootY) {
            if (this.rank[x] > this.rank[y]) {
                this.root[rootY] = rootX;
            } else if (this.rank[x] < this.rank[y]) {
                this.root[rootX] = rootY;
            } else {
                this.root[rootY] = rootX;
                this.rank[rootX]++;
            }
            this.count--;
        }
    }

    connected(x: number, y: number): boolean {
        return this.find(x) === this.find(y);
    }

    totalCounts(): number {
        return this.count;
    }
}

function earliestAcq(logs: number[][], n: number): number {
    logs.sort((a, b) => a[0] - b[0]);
    const disjointSet = new DisjointSetWithCount(n);

    for (const [time, x, y] of logs) {
        disjointSet.union(x, y);
        if (disjointSet.totalCounts() === 1) {
            return time;
        }
    }
    return -1;
}

File: solutions/c++/problems/[1101_Medium] The Earliest Moment When Everyone Become Friends.cpp
class DisjointSet {
    vector<int> root;
    vector<int> rank;
    unordered_set<int> activeRoots;
public:
    DisjointSet(int n) {
        root.reserve(n);
        rank = vector<int>(n, 0);
        for (int i = 0; i < n; i++) {
            root[i] = i;
            activeRoots.insert(i);
        }
    }

    void unionSet(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);
        if (rank[rootX] < rank[rootY]) {
            root[rootX] = rootY;
            activeRoots.erase(rootX);
        } else {
            root[rootY] = rootX;
            activeRoots.erase(rootY);
            if (rank[rootX] == rank[rootY]) rank[rootX]++;
        }
    }

    int find(int x) {
        if (root[x] == x) return x;
        return root[x] = find(root[x]);
    }

    bool connected(int x, int y) {
        return find(x) == find(y);
    }

    int getRootCounts() {
        return activeRoots.size();
    }
};

class Solution {
public:
    int earliestAcq(vector <vector<int>> &logs, int n) {
        sort(logs.begin(), logs.end(), [](auto &log1, auto &log2) {
            return log1[0] < log2[0];
        });
        auto ds = new DisjointSet(n);
        for (auto log: logs) {
            if (!ds->connected(log[1], log[2])) {
                ds->unionSet(log[1], log[2]);
                if (ds->getRootCounts() == 1) return log[0];
            }
        }
        return -1;
    }
};